homeassistant/components/asuswrt/sensor.py

```diff
@@ -1,184 +1,146 @@
-                _LOGGER.info("Reconnected to ASUS router for %s update", self.entity_id)
-        except OSError as err:
-            if not self._connect_error:
-                self._connect_error = True
-                _LOGGER.error(
-                    "Error connecting to ASUS router for %s update: %s",
-                    self.entity_id,
-                    err,
-                )
-
-
-class AsuswrtDevicesSensor(AsuswrtSensor):
-    """Representation of a asuswrt download speed sensor."""
-
-    _name = "Asuswrt Devices Connected"
-
-    async def async_update(self):
-        """Fetch new state data for the sensor."""
-        await super().async_update()
-        if self._devices:
-            self._state = len(self._devices)
-
-
-class AsuswrtRXSensor(AsuswrtSensor):
-    """Representation of a asuswrt download speed sensor."""
-
-    _name = "Asuswrt Download Speed"
-    _unit = DATA_RATE_MEGABITS_PER_SECOND
+    def icon(self) -> Optional[str]:
+        """Return the expected icon for the sensortype."""
+        if self in (_SensorTypes.UPLOAD, _SensorTypes.UPLOAD_SPEED):
+            return UPLOAD_ICON
+        if self in (_SensorTypes.DOWNLOAD, _SensorTypes.DOWNLOAD_SPEED):
+            return DOWNLOAD_ICON
 
     @property
-    def icon(self):
-        """Return the icon."""
-        return DOWNLOAD_ICON
+    def is_speed(self) -> bool:
+        """Return True if the type is an upload/download speed."""
+        return self in (_SensorTypes.UPLOAD_SPEED, _SensorTypes.DOWNLOAD_SPEED)
 
     @property
-    def unit_of_measurement(self):
-        """Return the unit of measurement."""
-        return self._unit
+    def is_size(self) -> bool:
+        """Return True if the type is the total upload/download size."""
+        return self in (_SensorTypes.UPLOAD, _SensorTypes.DOWNLOAD)
 
-    async def async_update(self):
-        """Fetch new state data for the sensor."""
-        await super().async_update()
-        if self._speed:
-            self._state = round(self._speed[0] / 125000, 2)
 
+async def async_setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up the asuswrt sensors."""
+    if discovery_info is None:
+        return
 
-class AsuswrtTXSensor(AsuswrtSensor):
-    """Representation of a asuswrt upload speed sensor."""
+    api: AsusWrt = hass.data[DATA_ASUSWRT]
 
-    _name = "Asuswrt Upload Speed"
-    _unit = DATA_RATE_MEGABITS_PER_SECOND
+    # Let's discover the valid sensor types.
+    sensors = [_SensorTypes(x) for x in discovery_info]
 
-    @property
-    def icon(self):
-        """Return the icon."""
-        return UPLOAD_ICON
+    data_handler = AsuswrtDataHandler(sensors, api)
+    coordinator = DataUpdateCoordinator(
+        hass,
+        _LOGGER,
+        name="sensor",
+        update_method=data_handler.update_data,
+        # Polling interval. Will only be polled if there are subscribers.
+        update_interval=timedelta(seconds=30),
+    )
 
```

Comment:
We should fetch data once with the coordinator before adding the entities.

Reply:
I had that initially, but I'm having a lot of trouble where the initialization seems to be stuck on this update. So I get an error message that it's taking 10, and 60 seconds and the setup is continuing without waiting longer. Also the sensors are not updated afterwards.

Using this implementation the sensors start at unavailable and eventually get the expected values, there are also no warnings that those updates cost too much time.

Reply:
My expectation is that this is due to a race issue as mentioned in an earlier comment, which has been fixed in the library, but not been released / updated in HA yet. So I'm unsure what the best way forward is.